src/gc.c

```diff
@@ -74,31 +73,24 @@ static void threadCallback(void* data) {
       RedisModule_UnblockClient(bc, NULL);
     }
     rm_free(task);
-    goto end;
-  }
-
-  if (!ret) {
+  } else if (!ret) {
     // The index was freed. There is no need to reschedule the task.
     // We need to free the task and the GC.
     RedisModule_Log(RSDummyContext, REDISMODULE_LOGLEVEL_DEBUG, "GC %p: Self-Terminating. Index was freed.", gc);
     gc->callbacks.onTerm(gc->gcCtx);
     if (bc) RedisModule_UnblockClient(bc, NULL);
     rm_free(task);
     rm_free(gc);
-    goto end;
-  }
-
-  RedisModule_ThreadSafeContextLock(ctx);
-  if (gc->timerID) {
-    gc->timerID = scheduleNext(task);
   } else {
-    RedisModule_Log(RSDummyContext, REDISMODULE_LOGLEVEL_DEBUG, "GC %p: Not scheduling next collection", gc);
-    rm_free(task);
+    RedisModule_ThreadSafeContextLock(RSDummyContext);
```

Comment:
why do we need the lock here?

Reply:
1. To read the timer ID, as it might get changed from a debug command and there is no other guard for that.
2. To create a new timer using the RM API we must hold the GIL